File: src/20-challenge/solution.py
type GiftListFixed = dict[str, dict[str, int]]


def fix_gift_list(received: list[str], expected: list[str]) -> GiftListFixed:
    gift_list_fixed: GiftListFixed = {
        "missing": {},
        "extra": {},
    }

    for gift in expected:
        try:
            gift_i: int = received.index(gift)
            received.pop(gift_i)
        except ValueError:
            gift_list_fixed["missing"].update(
                {gift: gift_list_fixed["missing"].get(gift, 0) + 1}
            )

    for gift in received:
        gift_list_fixed["extra"].update(
            {gift: gift_list_fixed["extra"].get(gift, 0) + 1}
        )

    return gift_list_fixed

File: src/20-challenge/solution.ts
interface GiftListFixed {
	missing: Record<string, number>
	extra: Record<string, number>
}

export default function fixGiftList(received: string[], expected: string[]): GiftListFixed {
	const giftListFixed: GiftListFixed = {
		missing: {},
		extra: {},
	}

	for (const gift of expected) {
		const giftI: number = received.indexOf(gift)

		if (~giftI) {
			received.splice(giftI, 1)
			continue
		}

		giftListFixed.missing[gift] = ~~giftListFixed.missing[gift] + 1
	}

	for (const gift of received) {
		giftListFixed.extra[gift] = ~~giftListFixed.extra[gift] + 1
	}

	return giftListFixed
}
